### PR TITLE
Async systems with sync clients.

### PR DESCRIPTION
**Pull request by felipefzdz**

### CODE REVIEW

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+Polling is by nature inefficient but could be a good solution when the technological stack doesn't allow bidirectional protocols like Web Sockets.
+
+### Hiding asynchronicity
+
+If we don't have control over those clients, we'll probably have to hide our async nature under some sync layer. That layer will implement a polling or pub/sub mechanism bounded by a timeout.
+
+## Learning by example
+
+In this series of posts we'll implement last strategy with a pub/sub mechanism. We'll add some essential complexities to our domain to make more juicy the exercise.
```
Comment (roberttaylor426) [inline _posts/2016-04-28-async-systems-with-sync-clients.md]: ...to make the exercise more juicy.

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      href: https://www.flickr.com/photos/barrasa8/2492558453
+tags:
+- distributed systems
+- akka
+- integration
+- architecture
+
+---
+
+As [Reactive Manifesto](http://www.reactivemanifesto.org/) says Reactive systems are: Responsive, Resilient, Elastic and Message Driven. The last principle often goes together with non-blocking async protocols. This style of communication "allows recipients to only consume resources while active, leading to less system overhead". This fits perfectly with new demands of efficiency derived from the elastic model of cloud providers. However not every system is async and integrating an async system with a sync client could be tricky.
```
Comment (roberttaylor426) [inline _posts/2016-04-28-async-systems-with-sync-clients.md]: As the Reactive Manifesto says...

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+
+## Integration Strategies
+
+We can fix the mismatch of communication styles through a couple of strategies.
+
+### Polling
+
+This involves work for both parties. First round trip of this protocol involves client sending some request to the server with an outcome of ACK/NACK. This is called fire and forget. Assuming HTTP, the server will return status code 202 (Accepted). The async process will eventually succeed or fail and that result will be exposed by the server in a different endpoint. Client will have to poll periodically that endpoint to figure out the status of the operation.
```
Comment (roberttaylor426) [inline _posts/2016-04-28-async-systems-with-sync-clients.md]: ...will have to periodically poll...

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+### Hiding asynchronicity
+
+If we don't have control over those clients, we'll probably have to hide our async nature under some sync layer. That layer will implement a polling or pub/sub mechanism bounded by a timeout.
+
+## Learning by example
+
+In this series of posts we'll implement last strategy with a pub/sub mechanism. We'll add some essential complexities to our domain to make more juicy the exercise.
+
+We'll be working on something similar to Craiglist, a website with classified advertisements. However our platform will have a focus on social (as everything nowadays). That means that a user can post an item into some group and/or to her followers. People can report dubious items and we will take seriously those reports as they're threats to our reputation. So much so that the authorities has direct access to an API that can take down an item immediately.
```
Comment (roberttaylor426) [inline _posts/2016-04-28-async-systems-with-sync-clients.md]: ...authorities have...

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+title: 'Async systems with sync clients'
+date: 2016-04-28 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/time.png
+    attribution:
+      text: Water time by barrasa8
+      href: https://www.flickr.com/photos/barrasa8/2492558453
+tags:
+- distributed systems
```
Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:13]: worth adding these as tags:
- sync
- async
- reactive or reactive systems

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+- akka
+- integration
+- architecture
+
+---
+
+As the [Reactive Manifesto](http://www.reactivemanifesto.org/) says Reactive systems are: Responsive, Resilient, Elastic and Message Driven. The last principle often goes together with non-blocking async protocols. This style of communication "allows recipients to only consume resources while active, leading to less system overhead". This fits perfectly with new demands of efficiency derived from the elastic model of cloud providers. However not every system is async and integrating an async system with a sync client could be tricky.
```
Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:20]: ...while staying active...

split the manifesto points into separate bullet points, so we bring attention of the readers to it

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+
+## Integration Strategies
+
+We can fix the mismatch of communication styles through a couple of strategies.
+
+### Polling
+
+This involves work for both parties. First round trip of this protocol involves client sending some request to the server with an outcome of ACK/NACK. This is called fire and forget. Assuming HTTP, the server will return status code 202 (Accepted). The async process will eventually succeed or fail and that result will be exposed by the server in a different endpoint. Client will have to periodically poll that endpoint to figure out the status of the operation.
```
Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:29]: ...via a different endpoint... ?

The Client ...

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+Polling is by nature inefficient but could be a good solution when the technological stack doesn't allow bidirectional protocols like Web Sockets.
+
+### Hiding asynchronicity
+
+If we don't have control over those clients, we'll probably have to hide our async nature under some sync layer. That layer will implement a polling or pub/sub mechanism bounded by a timeout.
```
Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:35]: 👍

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+## Learning by example
+
+In this series of posts we'll implement last strategy with a pub/sub mechanism. We'll add some essential complexities to our domain to make the exercise more juicy.
+
+We'll be working on something similar to Craiglist, a website with classified advertisements. However our platform will have a focus on social (as everything nowadays). That means that a user can post an item into some group and/or to her followers. People can report dubious items and we will take seriously those reports as they're threats to our reputation. So much so that the authorities have direct access to an API that can take down an item immediately.
```
Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:41]: focus on social ===> social focus or social theme ?

File: _posts/2016-04-28-async-systems-with-sync-clients.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+Our system is formed by several microservices based on Akka, using Kafka for inter-process communication. Police platform has only sync clients and they don't seem keen to implement a polling mechanism to verify that an item has been actually removed. We need to communicate with them synchronously and that's not negotiable. In next posts we'll see the details of our solution, but as an advancement, let's see the high level architecture. Don't worry if you don't understand everything yet.
```
Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:43]: The policing platform instead of Police platform or The Police platform ?

Comment (neomatrix369) [inline _posts/2016-04-28-async-systems-with-sync-clients.md:43]: In the next posts...

**Comment by neomatrix369** [comment]
Provide hot links for technologies or websites mentioned

**Comment by neomatrix369** [comment]
I was also wondering if using the terms Promise and Future in this post would be a good idea, up to you to decide. Because you already have started explaining the concepts in this post.

**Comment by neomatrix369** [comment]
Minor comment: use I/O and not IO throughout the post

**Comment by felipefzdz** [comment]
I'll talk more about Futures in further posts. This one is a high level introduction to the problem, so I'd prefer to keep it without a lot of implementation details.

**Comment by felipefzdz** [comment]
I couldn't find any reference to IO in this post :)

**Pull request merged**